src/wheel/wheelfile.py

```diff
@@ -1,169 +1,311 @@
+class WheelFile:
+    __slots__ = ('generator', 'root_is_purelib', '_mode', '_metadata', '_zip', '_data_path',
+                 '_dist_info_path', '_record_path', '_record_entries', '_exclude_archive_names')
+
+    # dist-info file names ignored for hash checking/recording
+    _exclude_filenames = ('RECORD', 'RECORD.jws', 'RECORD.p7s')
+    _default_hash_algorithm = 'sha256'
+
+    def __init__(self, path_or_fd: Union[str, PathLike, IO[bytes]], mode: str = 'r', *,
+                 metadata: Optional[WheelMetadata] = None, compression: int = ZIP_DEFLATED,
+                 generator: Optional[str] = None, root_is_purelib: bool = True):
+        if mode not in ('r', 'w'):
+            raise ValueError("mode must be either 'r' or 'w'")
+
+        if isinstance(path_or_fd, (str, PathLike)):
+            path_or_fd = Path(path_or_fd).open(mode + 'b')
+
+        if metadata is None:
+            if isinstance(path_or_fd, FileIO):
+                metadata = parse_filename(path_or_fd.name)
+            else:
+                raise WheelError('No file name or metadata provided')
+
+        self.generator = generator or 'Wheel {}'.format(wheel_version)
+        self.root_is_purelib = root_is_purelib
+        self._mode = mode
+        self._metadata = metadata
+        self._data_path = '{meta.name}-{meta.version}.data'.format(meta=self._metadata)
+        self._dist_info_path = '{meta.name}-{meta.version}.dist-info'.format(meta=self._metadata)
+        self._record_path = self._dist_info_path + '/RECORD'
+        self._exclude_archive_names = frozenset(self._dist_info_path + '/' + fname
+                                                for fname in self._exclude_filenames)
+        self._zip = ZipFile(path_or_fd, mode, compression=compression)
+        self._record_entries = OrderedDict()  # type: Dict[str, WheelRecordEntry]
 
-        self.dist_info_path = '{}.dist-info'.format(self.parsed_filename.group('namever'))
-        self.record_path = self.dist_info_path + '/RECORD'
-        self._file_hashes = OrderedDict()
-        self._file_sizes = {}
         if mode == 'r':
-            # Ignore RECORD and any embedded wheel signatures
-            self._file_hashes[self.record_path] = None, None
-            self._file_hashes[self.record_path + '.jws'] = None, None
-            self._file_hashes[self.record_path + '.p7s'] = None, None
+            self._read_record()
+
+    @property
+    def path(self) -> Optional[Path]:
+        return Path(self._zip.filename) if self._zip.filename else None
 
-            # Fill in the expected hashes by reading them from RECORD
+    @property
+    def mode(self) -> str:
+        return self._mode
+
+    @property
+    def metadata(self) -> WheelMetadata:
+        return self._metadata
+
+    def close(self) -> None:
+        try:
+            if self.mode == 'w':
+                self._write_wheelfile()
+                self._write_record()
```

Comment:
See comment below. Maybe there should be a check here that the user wrote a valid `METADATA` file, as it's the only other mandatory file in the wheel?

Maybe a default `METADATA` file, containing the name and version (see [here](https://packaging.python.org/specifications/core-metadata/), these and metadata version are the only mandatory fields) could be added if the user didn't add one expliicitly?

Reply:
I was thinking about adding a `write_metadata_file()` method but that's already being used for a method that writes a file to the `.dist-info` directory. Perhaps I should rename that to `write_dist_info_file()`?

Reply:
Bikeshedding on the name, I'd prefer `write_distinfo_file`.

As I say, I'm not sure how far you'd want to go in being responsible for formatting the metadata. And I don't see a lot of value in having both `write_metadata_file(content)` and `write_distinfo_file("METADATA", content)`, so if you're not going to take metadata as keywords, I don't think an extra method is worth it.

Validating on close that a metadata file is included *would* be worthwhile, though, IMO.

Reply:
> Bikeshedding on the name, I'd prefer write_distinfo_file.

I tend to agree because `write_metadata_file()` would suggest writing the actual `METADATA` file.

Reply:
The `close()` method now generates a minimum `METADATA` file if one doesn't exist already.

Reply:
Nice! That's an even better approach, thanks 🙂

Reply:
I've pushed the implementation for this now.